prov/sm2/src/sm2.h

```diff
@@ -335,14 +335,24 @@ static inline size_t sm2_pop_xfer_entry(struct sm2_ep *ep,
 		container_of(ep->util_ep.av, struct sm2_av, util_av);
 	struct sm2_mmap *map = &av->mmap;
 	struct sm2_region *self_region = sm2_mmap_ep_region(map, ep->gid);
+	*xfer_entry = NULL;
 
 	if (smr_freestack_isempty(sm2_freestack(self_region))) {
 		sm2_progress_recv(ep);
+		self_region = sm2_mmap_ep_region(map, ep->gid);
 		if (smr_freestack_isempty(sm2_freestack(self_region)))
 			return -FI_EAGAIN;
 	}
-
 	*xfer_entry = smr_freestack_pop(sm2_freestack(self_region));
+#if ENABLE_DEBUG
+	{
+		assert(sm2_freestack(self_region)->free >= 0);
+		assert((char *) *xfer_entry >=
+		       (char *) sm2_freestack(self_region));
+		assert((char *) *xfer_entry <
+		       (char *) sm2_mmap_ep_region(map, ep->gid + 1));
```

Comment:
So, `sm2_mmap_ep_region(map, ep->gid + 1)` is the beginning of the region for the next ep (if there is)? Does that mean the freestack is the only part of each region. This check does not look intuitive to me.

Reply:
> So, `sm2_mmap_ep_region(map, ep->gid + 1)` is the beginning of the region for the next ep (if there is)?

Correct.  As I said above this is a crude check.  It's not a tight bound on the freestack but should be enough to find runaway problems.

I think it might actually be the last section of the region, but I'm not 100% sure.

Reply:
Also, if you prefer I can just remove these asserts

Reply:
No it doesn't hurt to check more things in DEBUG build, it would be good if you can add a comment here to explain